claat/serve.go

```diff
@@ -0,0 +1,221 @@
+	"polymer":             "Polymer/polymer#standard-layer",
+	"url-search-params":   "WebReflection/url-search-params#^0.5.0",
+}
+
+var depsDir = "bower_components"
+var elemDir = "elements"
+var elemFile = elemDir + "/codelab.html"
+
+const codelabElem = `
+<link rel="import" href="../bower_components/google-codelab-elements/google-codelab-elements.html">
+<link rel="import" href="../bower_components/google-youtube/google-youtube.html">
+<link rel="import" href="../bower_components/iron-ajax/iron-ajax.html">
+`
+
+// cmdServe is the "claat serve ..." subcommand.
+func cmdServe() {
+	err := mkdirIfNec(depsDir)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	// If we don't have the required dependencies, go get 'em.
+	for k, v := range deps {
+		if _, err := os.Stat(depsDir + "/" + k); os.IsNotExist(err) {
+			s := strings.Split(v, "#")
+			url := "https://github.com/" + s[0] + "/archive/master.zip"
+			archFile := depsDir + "/master.zip"
+			destDir := depsDir + "/" + k
+			fmt.Println("Downloading " + url)
+			err := DownloadFile(archFile, url)
+			if err != nil {
+				fmt.Println(err)
+				continue
+			}
+			err = Unzip(archFile, depsDir)
+			os.Rename(destDir+"-master", destDir)
+			os.Remove(archFile)
+			if err != nil {
+				fmt.Println(err)
+				continue
+			}
+		}
+	}
+	err = mkdirIfNec(elemDir)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	if _, err := os.Stat(elemFile); os.IsNotExist(err) {
+		f, err := os.Create(elemFile)
+		if err != nil {
+			log.Fatal("Cannot create file", err)
+		}
+		defer f.Close()
+		fmt.Fprintf(f, codelabElem)
+	}
+	fmt.Println("Dependencies installed.")
+
+	port := "9090"
+	http.Handle("/", http.FileServer(http.Dir(".")))
+	fmt.Println("Serving on localhost:" + port + ", opening browser tab now...")
+	openBrowser("http://127.0.0.1:" + port)
+	err = http.ListenAndServe(":"+port, nil) // set listen port
+	if err != nil {
+		log.Fatal("ListenAndServe: ", err)
+	}
+}
+
+// DownloadFile will download a url to a local file. It's efficient because it will
+// write as it downloads and not load the whole file into memory.
+func DownloadFile(filepath string, url string) error {
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Unzip a file to a dest directory.
+func Unzip(src, dest string) error {
+	r, err := zip.OpenReader(src)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Close(); err != nil {
+			panic(err)
+		}
+	}()
+	os.MkdirAll(dest, 0755)
+	// Closure to address file descriptors issue with all the deferred .Close() methods
+	extractAndWriteFile := func(f *zip.File) error {
+		rc, err := f.Open()
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := rc.Close(); err != nil {
+				panic(err)
```

Comment:
Can we just ingnore rc.Close() errors? It doesn't seem that much important. Maybe I'm missing something.

Reply:
agreed